### ISSUE TITLE
Code lens to display type signature for local bindings in where clauses

### ISSUE THREAD

**Issue by July541** [haskell/haskell-language-server#2943]
Reminded by https://github.com/haskell/haskell-language-server/issues/2019#issuecomment-1144528250, I think it would be great to have a configurable plugin to display a signature code lens for local bindings in where clauses.

Please let me know if there has more places need code lenses.

~~Edited: I'll try to pick bindings from [`RenamedSource`](https://hackage.haskell.org/package/ghc-9.0.2/docs/GHC.html#t:RenamedSource), please let me know if there is a more efficient way.~~ Should be TypeCheckedSource.

**Comment by michaelpj** [comment]
> configurable

Configuration is bad! I say have it be enabled whenever the main type signature lens is.

**Comment by July541** [comment]
> Configuration is bad! I say have it be enabled whenever the main type signature lens is.

Could you please explain why configuration is bad? Ideally I want to control everything displayed on my screen by category...

**Comment by michaelpj** [comment]
Because it means more things for users to think about, a bigger settings page, and more things for us to support. I think if we can, it's best only to introduce configuration in cases where there's a real split in what people want *and* people are annoyed if they don't get the one they want.

**Comment by michaelpj** [comment]
I think that's reasonable. Especially since I'd hope the code is mostly there already!

**Comment by uhbif19** [comment]
Does not it already show type signatures?

![image](https://user-images.githubusercontent.com/300919/174140856-9a0f2724-c511-4d14-bc37-cee6233087ce.png)

**Comment by July541** [comment]
> Does not it already show type signatures?

Yes, but it requires users hover on it, I'm going to display code lens directly like this:

![image](https://user-images.githubusercontent.com/16057632/174243407-d592504a-a127-4330-98a6-1c6fd8410c1f.png)

**Comment by uhbif19** [comment]
@July541 Now I got it, I misunderstood that "code lens" is, I thought lens is a bubble. 

That is cool. But why this issue is concerned only on where clauses? Should be cool to see this in let clauses too. Now I do not get them.

![image](https://user-images.githubusercontent.com/300919/174396533-c9a566e3-3eea-483d-86ca-93dd332c2b71.png)

**Comment by July541** [comment]
I initially want to display sigs for let bindings by inlay hints, see https://github.com/haskell/haskell-language-server/issues/2938, but now I think it's not bad to show code lens too, maybe it is worth trying after #2966 :)

**Comment by michaelpj** [comment]
I think doing a code lens is good: code lenses fit better for places where you can put signatures on their own lines, which is probably better in general.
